test/modifiers.c

```diff
@@ -151,10 +157,67 @@ test_modmap_none(void)
     xkb_context_unref(context);
 }
 
+static void
+test_real_mods(void)
+{
+    struct xkb_context *context = test_get_context(0);
+    struct xkb_keymap *keymap;
+    const struct xkb_key *key;
+    xkb_keycode_t keycode;
+
+    keymap = test_compile_file(context, "keymaps/real-modifiers.xkb");
+    assert(keymap);
+
+    keycode = xkb_keymap_key_by_name(keymap, "AD01");
+    assert(keycode != XKB_KEYCODE_INVALID);
+    key = XkbKey(keymap, keycode);
+    assert(key->modmap == Mod5Mask);
+    assert(key->vmodmap == Mod10Mask << 1);
+
+    keycode = xkb_keymap_key_by_name(keymap, "AD02");
+    assert(keycode != XKB_KEYCODE_INVALID);
+    key = XkbKey(keymap, keycode);
+    assert(key->modmap == Mod6Mask);
+    assert(key->vmodmap == Mod10Mask << 2);
+
+    assert(test_key_seq(keymap,
+                        KEY_O, BOTH, XKB_KEY_a, NEXT,
+                        KEY_Q, DOWN, XKB_KEY_q, NEXT,
+                        KEY_O, BOTH, XKB_KEY_c, NEXT,
+                        KEY_Q, UP,   XKB_KEY_q, NEXT,
+                        KEY_W, DOWN, XKB_KEY_w, NEXT,
+                        KEY_O, BOTH, XKB_KEY_d, NEXT,
+                        KEY_W, UP,   XKB_KEY_w, NEXT,
+                        KEY_E, DOWN, XKB_KEY_e, NEXT,
+                        KEY_O, BOTH, XKB_KEY_f, NEXT,
+                        KEY_E, UP,   XKB_KEY_e, NEXT,
+                        KEY_R, DOWN, XKB_KEY_r, NEXT,
+                        KEY_O, BOTH, XKB_KEY_g, NEXT,
+                        KEY_R, UP,   XKB_KEY_r, NEXT,
+                        KEY_T, DOWN, XKB_KEY_t, NEXT,
+                        KEY_O, BOTH, XKB_KEY_h, NEXT,
+                        KEY_T, UP,   XKB_KEY_t, NEXT,
+                        KEY_Y, DOWN, XKB_KEY_y, NEXT,
+                        KEY_O, BOTH, XKB_KEY_i, NEXT,
+                        KEY_Y, UP,   XKB_KEY_y, NEXT,
+                        KEY_U, DOWN, XKB_KEY_u, NEXT,
+                        KEY_O, BOTH, XKB_KEY_j, NEXT,
+                        KEY_U, UP,   XKB_KEY_u, NEXT,
+                        KEY_Q, DOWN, XKB_KEY_q, NEXT,
+                        KEY_W, DOWN, XKB_KEY_w, NEXT,
+                        KEY_O, BOTH, XKB_KEY_j, NEXT,
+                        KEY_W, UP,   XKB_KEY_w, NEXT,
+                        KEY_Q, UP,   XKB_KEY_q, FINISH));
```

Comment:
staring at `test_key_seq` - it doesn't actually test any of the modifiers being set here. which arguably here is quite important ;)

Reply:
Well we do: the various `KEY_Q..KEY_I` are mapped to virtual modifiers through the keysyms `q..i` (now `Q..I`). The interesting result is what we get from `KEY_O` (`<AD09>`), which uses our special key type `SOOO_MANY_LEVELS`.

Reply:
*audible click sound in the background* 

ah, now i get it, thanks!